### PR TITLE
storcon: fix case where we might fail to send compute notifications after two opposite migrations 

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

If we migrate A->B, then B->A, and the notification of A->B fails, then we might have retained state that makes us think "A" is the last state we sent to the compute hook, whereas when we migrate B->A we should really be sending a fresh notification in case our earlier failed notification has actually mutated the remote compute config.

Closes: #9417 

## Summary of changes

- Add a reproducer for the bug (`test_storage_controller_compute_hook_revert`)
- Refactor compute hook code to represent remote state with `ComputeRemoteState` which stores a boolean for whether the compute has fully applied the change as well as the request that the compute accepted.
- The actual bug fix: after sending a compute notification, if we got a 423 response then update our ComputeRemoteState to reflect that we have mutated the remote state.  This way, when we later try and notify for our historic location, we will properly see that as a change and send the notification.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Review by mattpodraza — APPROVED** [review]
Thanks!

File: test_runner/regress/test_storage_controller.py [hunk 1]
```diff
@@ -719,6 +720,113 @@ def logged_stuck_again():
     env.storage_controller.consistency_check()
 
 
+def test_storage_controller_compute_hook_revert(
```
Comment (VladLazar) [inline test_runner/regress/test_storage_controller.py:724]: nit: we can run this only on the default pg version

File: storage_controller/src/compute_hook.rs [hunk 2]
```diff
@@ -38,7 +38,22 @@ struct ShardedComputeHookTenant {
...
+/// Represents our knowledge of the compute's state: we can update this when we get a
+/// response from a notify API call, which tells us what has been applied.
+///
+/// Should be wrapped in an Option<>, as we cannot always know the remote state.
+#[derive(PartialEq, Eq, Debug)]
+struct ComputeRemoteState {
+    // The request body which was acked by the compute
+    request: ComputeHookNotifyRequest,
+
+    // Whether the compute indicated that the state was applied to running computes, or just
```
[the remark this answers is not in the document]
Comment (VladLazar) [reply storage_controller/src/compute_hook.rs]: https://github.com/neondatabase/neon/pull/9435/commits/cd2775c409ad9f2f4957451846efede5e5e6d0b1

File: test_runner/regress/test_storage_controller.py (hunk 1 above)
Comment (VladLazar) [reply test_runner/regress/test_storage_controller.py:724]: https://github.com/neondatabase/neon/pull/9435/commits/96f28069c00016e8457fd8072216fcc363f20f1f

**Review by VladLazar — APPROVED** [review]

**Pull request merged**